ksqldb-engine/src/main/java/io/confluent/ksql/query/QueryBuilder.java

```diff
@@ -650,4 +671,131 @@ private static void updateListProperty(
     valueList.add(value);
     properties.put(key, valueList);
   }
+
+  private static class TransientQuerySinkProcessor implements
+      Processor<Object, GenericRow, Void, Void> {
+
+    private final TransientQueryQueue queue;
+    private final Optional<ImmutableMap<TopicPartition, Long>> endOffsets;
+    private final ConcurrentHashSet<TopicPartition> donePartitions;
+    private ProcessorContext<Void, Void> context;
+
+    TransientQuerySinkProcessor(
+        final TransientQueryQueue queue,
+        final Optional<ImmutableMap<TopicPartition, Long>> endOffsets,
+        final ConcurrentHashSet<TopicPartition> donePartitions) {
+
+      this.queue = queue;
+      this.endOffsets = endOffsets;
+      this.donePartitions = donePartitions;
+    }
+
+    @Override
+    public void init(final ProcessorContext<Void, Void> context) {
+      Processor.super.init(context);
+      this.context = context;
+      if (endOffsets.isPresent()) {
+        // registering a punctuation here is the easiest way to check
+        // if we're complete even if we don't see all records.
+        // The in-line check in process() is faster,
+        // but some records may be filtered out before reaching us.
+        context.schedule(
+            Duration.ofMillis(100),
+            PunctuationType.WALL_CLOCK_TIME,
+            timestamp -> checkCompletion()
+        );
+      }
+    }
+
+    @Override
+    public void process(final Record<Object, GenericRow> record) {
+      final Optional<TopicPartition> topicPartition =
+          context
+              .recordMetadata()
+              .map(m -> new TopicPartition(m.topic(), m.partition()));
+
+      final boolean alreadyDone =
+          topicPartition.isPresent() && donePartitions.contains(topicPartition.get());
+
+      // We ignore null values in streams as invalid data and drop the record.
+      // If we're already done with the current partition, we just drop the record.
+      if (record.value() != null && !alreadyDone) {
+        queue.acceptRow(null, record.value());
+      }
+
+      if (topicPartition.isPresent()) {
+        checkCompletion(topicPartition.get());
+      }
+    }
+
+    private void checkCompletion(final TopicPartition topicPartition) {
```

Comment:
I found it just a bit confusing having 3 different versions of checkCompletion.  Maybe it makes sense to call this something like `checkCompletionPartition` the next `checkCompletionAllPartitions`?

If you had `checkCompletion()` call this method, you could inline the body of the third method here.  Is the idea that you want a consistent snapshot of currentPositions?  There is only one thread acting on everything, right?

Reply:
Roger that. I kind of phoned it in on naming these methods. I'll fix it.

It's handy to keep the third method so that we don't have to separately call `currentPositions` before invoking the single-partition version. Maybe you can take another look after I push my current batch of updates and see if you think it's ok.